coderd/autobuild/lifecycle_executor.go

```diff
@@ -177,6 +178,15 @@ func (e *Executor) runOnce(t time.Time) Stats {
 				err := e.db.InTx(func(tx database.Store) error {
 					var err error
 
+					ok, err := tx.TryAcquireLock(e.ctx, database.GenLockID(fmt.Sprintf("lifecycle-executor:%s", wsID)))
+					if err != nil {
+						return xerrors.Errorf("try acquire lifecycle executor lock: %w", err)
```

Comment:
Should we handle context cancelled and return nil? Not sure it matters, but we'll end up assigning `stats.Errors[wsID] = err`, granted the server is most likely shutting down so might not matter.

Reply:
Would this be covered by the higher up `case <-e.ctx.Done():`?

https://github.com/coder/coder/blob/c45817bc62a3e408310fab7818876ee7e546a4a4/coderd/autobuild/lifecycle_executor.go#L105-L117

Reply:
I realize that for what I'm proposing, the "problem" existed in the existing code as well and not something introduced in this PR.

To answer your question, I think yes and no. It will at least log a bunch of `log.Error(e.ctx, "failed to transition workspace", slog.Error(err))` (for each workspace that's being run in parallel), as well as potentially send stats on the channel in the code you referenced (the select is racy and it could pick either done or send stats as long as there's a listener).

Essentially if we decide to handle context cancelled it should probably be around line 402.

Reply:
Have now added a check so any context canceled errors shouldn't be included in logs and stats.

https://github.com/coder/coder/blob/46fa216a243247bd84903ed65a39c7ef09979516/coderd/autobuild/lifecycle_executor.go#L402-L410

Reply:
Nice find!